### PR TITLE
Refactor xpt2046 to be a touchscreen platform

### PR DESCRIPTION
**Pull request by nielsnl68**
** This a new PR because i messed up my older attempt. **

# What does this implement/fix?

Changing the way the xpt2046 touchscreen driver is configurated in the yaml file.
Instead of a standalone (top level) component it is now a platform for the touchscreen component.
Now it knows on which display-page on your and handles display rotation as well.
With some small changes here and there the properties stay almost the same.

`irq_pin`  is changed into `interupt_pin`
`on_state` => `on_touch`
`x_dim` and `y_dim` are not needed anymore.

the xpt2046 binarysensor is not removed and the touchscreen binarysensor should be used instead.

The old method is fully removed. howhever there should be a error warning to expain that you should use it as platform of the touchscreen.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other:
All existing yaml keys are still vailed and for the binarysensors the platform need to be updated to `touchscreen` instead of `xtp2046`

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2259

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
touchscreen:
  - platform: xpt2046
    display: my_display
    id: my_toucher
    cs_pin: 14
    #interupt_pin: 16
    report_interval: 50ms
    threshold: 400
    calibration_x_min: 3860
    calibration_x_max: 280
    calibration_y_min: 340
    calibration_y_max: 3860
    #swap_x_y: true

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/xpt2046/touchscreen.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        var.set_calibration(
+            config[CONF_CALIBRATION_X_MIN],
+            config[CONF_CALIBRATION_X_MAX],
+            config[CONF_CALIBRATION_Y_MIN],
+            config[CONF_CALIBRATION_Y_MAX],
+        )
+    )
+
+    if CONF_SWAP_X_Y in config:
+        cg.add(var.set_swap_x_y(config[CONF_SWAP_X_Y]))
```
Comment (jesserockz) [inline esphome/components/xpt2046/touchscreen.py]: As there is a default, its always in config
```suggestion
    cg.add(var.set_swap_x_y(config[CONF_SWAP_X_Y]))
```

**Review by jesserockz — APPROVED** [review]
One minor thing

File: esphome/components/xpt2046/touchscreen.py (hunk 1 above)
Comment (nielsnl68) [reply esphome/components/xpt2046/touchscreen.py]: Good catch, thanks for pointing this out.

**Comment by nielsnl68** [comment]
Sorry @Syndlex, to bother you by above message, i have to many branches i'm working on. that i missed that i changed something in the wrong branch. The change is already restored back.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**